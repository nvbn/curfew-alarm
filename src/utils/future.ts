/**
 * Abstraction around value that might not be yet available,
 * can be used with hooks when the values is produce asynchronously.
 * */

export const FUTURE_NOT_READY = "__FUTURE_NOT_READY__";

export type Future<T> = T | typeof FUTURE_NOT_READY;

export const isReady = <T>(val: Future<T>): val is T =>
  val !== FUTURE_NOT_READY;

export const futureMap = <T, Q>(v: Future<T>, fn: (v: T) => Q): Future<Q> =>
  isReady(v) ? fn(v) : v;
