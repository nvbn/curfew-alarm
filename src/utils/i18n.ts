import i18n, { TranslateOptions } from "i18n-js";

import { DEFAULT } from "../translations";

type Copy = keyof typeof DEFAULT;

type Scope = Copy | Copy[];

/**
 * i18n-js.t with typescript tags validation.
 */
const t = (scope: Scope, options?: TranslateOptions): string =>
  i18n.t(scope, options);

export default {
  ...i18n,
  t,
};
