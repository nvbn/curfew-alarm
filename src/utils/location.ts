import { INetwork, NETWORK_TYPE_WIFI } from "../dependencies/Network";

/**
 * When a user connected to wifi treated as they are at home.
 */
export const isAtHome = async (network: INetwork): Promise<boolean> => {
  const status = await network.getNetworkStateAsync();

  return status.type === NETWORK_TYPE_WIFI;
};
