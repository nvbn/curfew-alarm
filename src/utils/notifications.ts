import IConstants from "../dependencies/IConstants";
import INotifications, {
  NOTIFICATION_PERMISSIONS_DENIED,
  NOTIFICATIONS_PERMISSIONS_GRANTED,
  NOTIFICATIONS_PERMISSIONS_UNDETERMINED,
} from "../dependencies/INotifications";
import IPlatform, { PLATFORM_OS_ANDROID } from "../dependencies/IPlatform";

/**
 * Schedules local push notification.
 */
export const sendNotification = (
  notifications: INotifications,
  title: string,
  body: string,
): Promise<string> =>
  notifications.scheduleNotificationAsync({
    content: {
      title,
      body,
    },
    trigger: null,
  });

export const REGISTER_NOT_DEVICE = "REGISTER_NOT_DEVICE";

export const REGISTER_DENIED = "REGISTER_DECLINED";

export const REGISTER_OK = "REGISTER_OK";

type RegisterStatus = {
  status:
    | typeof REGISTER_NOT_DEVICE
    | typeof REGISTER_DENIED
    | typeof REGISTER_OK;
  token?: string;
};

/**
 * Registers app as push notifications sender / handler.
 */
export const registerForPushNotifications = async (
  { isDevice }: IConstants,
  { OS }: IPlatform,
  notifications: INotifications,
  reRequestPermissions: boolean,
): Promise<RegisterStatus> => {
  if (!isDevice) {
    return { status: REGISTER_NOT_DEVICE };
  }

  let status = (await notifications.getPermissionsAsync()).status;

  if (
    (reRequestPermissions && status === NOTIFICATION_PERMISSIONS_DENIED) ||
    status === NOTIFICATIONS_PERMISSIONS_UNDETERMINED
  ) {
    status = (await notifications.requestPermissionsAsync()).status;
  }

  if (status !== NOTIFICATIONS_PERMISSIONS_GRANTED) {
    return { status: REGISTER_DENIED };
  }

  const token = (await notifications.getExpoPushTokenAsync()).data;

  if (OS === PLATFORM_OS_ANDROID) {
    await notifications.setNotificationChannelAsync("default", {
      name: "default",
      importance: 5, // default
    });
  }

  return {
    status: REGISTER_OK,
    token,
  };
};
