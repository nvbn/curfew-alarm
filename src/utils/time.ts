import { isValidNumber } from "./numbers";

/**
 * Internal abstraction for storing time, it was just more convenient to have it.
 */
export type Time = {
  hour: number;
  minute: number;
};

/**
 * Time to minute from the start of the day.
 */
export const toMinutes = ({ hour, minute }: Time): number => hour * 60 + minute;

/**
 * Converts js Date to internal Time.
 */
export const dateToTime = (date: Date): Time => ({
  hour: date.getHours(),
  minute: date.getMinutes(),
});

const WE_DONT_CARE_ABOUT_THE_ACTUAL_DATE = "2021-01-10T00:00:00";

/**
 * Converts internal Time to js Date.
 */
export const timeToDate = ({ hour, minute }: Time): Date => {
  const date = new Date(WE_DONT_CARE_ABOUT_THE_ACTUAL_DATE);
  date.setHours(hour);
  date.setMinutes(minute);

  return date;
};

/**
 * Formats internal time in 24h form.
 */
export const formatTime = ({ hour, minute }: Time): string =>
  `${hour.toString().padStart(2, "0")}:${minute.toString().padStart(2, "0")}`;

/**
 * Returns true if passed object is a valid Time.
 */
// eslint-disable-next-line
export const isValidTime = (value: any): value is Time => {
  if (!value) {
    return false;
  }

  if (!isValidNumber(value.hour) || value.hour > 23 || value.hour < 0) {
    return false;
  }

  return !(
    !isValidNumber(value.minute) ||
    value.minute > 59 ||
    value.minute < 0
  );
};
