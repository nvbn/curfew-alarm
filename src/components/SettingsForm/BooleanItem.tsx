import React from "react";
import { Switch, Text, TouchableWithoutFeedback, View } from "react-native";

import { isReady } from "../../utils/future";
import styles from "./styles";
import { BooleanItemProps } from "./types";

/**
 * Checkbox/switch settings item for boolean values.
 */
const BooleanItem = ({
  id,
  title,
  value,
  onChange,
}: BooleanItemProps): JSX.Element => (
  <TouchableWithoutFeedback
    testID={`boolean-item-touchable-${id}`}
    onPress={() => onChange(!value)}
  >
    <View style={[styles.item, styles.booleanItem]}>
      <Text style={styles.title}>{title}</Text>
      {isReady(value) ? (
        <Switch
          value={value}
          onValueChange={onChange}
          testID={`boolean-item-switch-${id}`}
        />
      ) : (
        <Switch disabled />
      )}
    </View>
  </TouchableWithoutFeedback>
);

export default BooleanItem;
