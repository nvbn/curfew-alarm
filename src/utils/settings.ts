import IPersistentStorage from "../dependencies/IPersistentStorage";
import { isValidNumber } from "./numbers";
import { isValidTime, Time } from "./time";

export type Settings = {
  curfewStart: Time;
  curfewEnd: Time;

  minutesToGoHome: number;
};

// exported only for tests
export const DEFAULT_SETTINGS: Settings = {
  curfewStart: { hour: 21, minute: 0 },
  curfewEnd: { hour: 4, minute: 30 },
  minutesToGoHome: 30,
};

// exported only for tests
export const STORAGE_KEY = "SETTINGS";

const parseSettings = (serialised: string | null): Settings => {
  let settings = DEFAULT_SETTINGS;
  if (serialised !== null) {
    try {
      settings = { ...settings, ...JSON.parse(serialised) };
    } catch (e) {
      console.warn("unable to decode settings", e);
    }
  }

  if (!isValidTime(settings.curfewStart)) {
    console.warn("incorrect value in curfewStart", settings.curfewStart);
    settings.curfewStart = DEFAULT_SETTINGS.curfewStart;
  }

  if (!isValidTime(settings.curfewEnd)) {
    console.warn("incorrect value in curfewEnd", settings.curfewEnd);
    settings.curfewEnd = DEFAULT_SETTINGS.curfewEnd;
  }

  if (
    !isValidNumber(settings.minutesToGoHome) ||
    settings.minutesToGoHome < 0
  ) {
    console.warn(
      "incorrect value in minutesToGoHome",
      settings.minutesToGoHome,
    );
    settings.minutesToGoHome = DEFAULT_SETTINGS.minutesToGoHome;
  }

  return settings;
};

/**
 * Reads settings from persistent storage, returns default settings when
 * it's not possible, and replaces invalid keys with default values.
 */
export const getSettings = async (
  storage: IPersistentStorage,
): Promise<Settings> => {
  try {
    const serialised = await storage.getItem(STORAGE_KEY);

    return parseSettings(serialised);
  } catch (e) {
    console.warn("unable to read settings", e);

    return DEFAULT_SETTINGS;
  }
};

/**
 * Saves settings in persisten storage.
 */
export const updateSettings = (
  storage: IPersistentStorage,
  settings: Settings,
): Promise<void> => storage.setItem(STORAGE_KEY, JSON.stringify(settings));
