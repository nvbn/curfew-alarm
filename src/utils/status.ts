import { Time, toMinutes } from "./time";

export const STATUS_FINE = "FINE";

export const STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME =
  "GO_HOME_WHEN_TIME_TO_GO_HOME";

export const STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME =
  "STAY_AT_HOME_WHEN_TIME_TO_GO_HOME";

export const STATUS_STAY_AT_HOME_WHEN_CURFEW = "STAY_AT_HOME_WHEN_CURFEW";

export const STATUS_GO_HOME_WHEN_CURFEW = "GO_HOME_WHEN_CURFEW";

export type Status =
  | typeof STATUS_FINE
  | typeof STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME
  | typeof STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME
  | typeof STATUS_STAY_AT_HOME_WHEN_CURFEW
  | typeof STATUS_GO_HOME_WHEN_CURFEW;

const MAX_MINUTE = 24 * 60;

/**
 * Determines users status / what a user needs to do at the moment
 * based on time and configurations.
 */
export const getStatus = (
  enabled: boolean,
  isAtHome: boolean,
  currentTime: Time,
  curfewStart: Time,
  curfewEnd: Time,
  minutesToGoHome: number,
): Status => {
  if (!enabled) {
    return STATUS_FINE;
  }

  const currentMinute = toMinutes(currentTime);
  const curfewStartMinute = toMinutes(curfewStart);
  const curfewEndMinute = toMinutes(curfewEnd);
  const goHomeStartMinute = curfewStartMinute - minutesToGoHome;

  if (
    currentMinute >= curfewStartMinute &&
    ((curfewEndMinute < curfewStartMinute && currentMinute <= MAX_MINUTE) ||
      currentMinute <= curfewEndMinute)
  ) {
    return isAtHome
      ? STATUS_STAY_AT_HOME_WHEN_CURFEW
      : STATUS_GO_HOME_WHEN_CURFEW;
  } else if (currentMinute >= goHomeStartMinute) {
    return isAtHome
      ? STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME
      : STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME;
  } else {
    return STATUS_FINE;
  }
};
