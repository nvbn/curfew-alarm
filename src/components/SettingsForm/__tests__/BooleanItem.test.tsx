/* eslint-disable @typescript-eslint/no-empty-function */
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";

import { FUTURE_NOT_READY } from "../../../utils/future";
import BooleanItem from "../BooleanItem";

describe("<BooleanItem />", () => {
  test("can be rendered", () => {
    const rendered = render(
      <BooleanItem
        id="test"
        title="Some toggle"
        type="boolean"
        value={false}
        onChange={() => {}}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  test("can be rendered when value isn't ready", () => {
    const rendered = render(
      <BooleanItem
        id="test"
        title="Some toggle"
        type="boolean"
        value={FUTURE_NOT_READY}
        onChange={() => {}}
      />,
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  test("fires onChange when switch touched", () => {
    const onChange = jest.fn();

    const { getByTestId } = render(
      <BooleanItem
        id="test"
        title="Some toggle"
        type="boolean"
        value={true}
        onChange={onChange}
      />,
    );

    const toggle = getByTestId("boolean-item-switch-test");
    fireEvent(toggle, "valueChange", false);

    expect(onChange.mock.calls).toEqual([[false]]);
  });

  test("fires onChange when the whole item touched", () => {
    const onChange = jest.fn();

    const { getByTestId } = render(
      <BooleanItem
        id="test"
        title="Some toggle"
        type="boolean"
        value={true}
        onChange={onChange}
      />,
    );

    const toggle = getByTestId("boolean-item-touchable-test");
    fireEvent.press(toggle);

    expect(onChange.mock.calls).toEqual([[false]]);
  });
});
