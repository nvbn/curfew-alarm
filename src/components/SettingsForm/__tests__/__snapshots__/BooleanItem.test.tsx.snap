// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BooleanItem /> can be rendered 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderBottomColor": "#fafafa",
        "borderBottomWidth": 1,
        "marginHorizontal": 16,
        "marginVertical": 8,
        "paddingHorizontal": 10,
        "paddingVertical": 20,
      },
      Object {
        "alignItems": "flex-start",
      },
    ]
  }
  testID="boolean-item-touchable-test"
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    Some toggle
  </Text>
  <RCTSwitch
    accessibilityRole="switch"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 31,
        "width": 51,
      }
    }
    testID="boolean-item-switch-test"
    value={false}
  />
</View>
`;

exports[`<BooleanItem /> can be rendered when value isn't ready 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderBottomColor": "#fafafa",
        "borderBottomWidth": 1,
        "marginHorizontal": 16,
        "marginVertical": 8,
        "paddingHorizontal": 10,
        "paddingVertical": 20,
      },
      Object {
        "alignItems": "flex-start",
      },
    ]
  }
  testID="boolean-item-touchable-test"
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    Some toggle
  </Text>
</View>
`;
